src/Formats/ProtobufWriter.cpp

```diff
@@ -123,21 +124,38 @@ namespace
 
 // SimpleWriter is an utility class to serialize protobufs.
 // Knows nothing about protobuf schemas, just provides useful functions to serialize data.
-ProtobufWriter::SimpleWriter::SimpleWriter(WriteBuffer & out_) : out(out_), current_piece_start(0), num_bytes_skipped(0)
+ProtobufWriter::SimpleWriter::SimpleWriter(WriteBuffer & out_, const bool single_message_mode_)
+    : out(out_)
+    , current_piece_start(0)
+    , num_bytes_skipped(0)
+    , produce_length_delimiters(!single_message_mode_)
+    // normally for single_message_mode we forbid outputting more than one row
+    // (it would lead to malformed protobuf message), but Kafka/Rabbit push every row
+    // in a separate message, so it's valid case there and should be allowed.
+    , allow_several_messages(!single_message_mode_ || out.producesIsolatedRows())
+    , row_was_send(false)
 {
 }
 
 ProtobufWriter::SimpleWriter::~SimpleWriter() = default;
 
 void ProtobufWriter::SimpleWriter::startMessage()
 {
+    if (!allow_several_messages && row_was_send)
```

Comment:
This looks too complex. `SimpleWriter` should be simple ;)
These two new member variables don't seem to be necessary at all.
Just add parameter `bool length_delimited` to `SimpleWriter::endMessage()` and that should be enough.

Reply:
But it we will write 2 protobuf messages one after another, it will not be readable. So i wanted to forbid attempts to produce more than single row for non-Kafka case.

Of course i can remove those checks... But we will be silently create protobuf which nobody can read..

Reply:
`SimpleWriter` is an internal class, so it's mostly the caller's responsibility to use it correctly. For example, there is no protection from calling `endMessage()` without `startMessage()`.